v2/pkg/engine/datasource/grpc_datasource/grpc_datasource_test.go

```diff
@@ -0,0 +1,1054 @@
+						Response: "QueryCategoriesByKindResponse",
+					},
+					"filterCategories": {
+						RPC:      "QueryFilterCategories",
+						Request:  "QueryFilterCategoriesRequest",
+						Response: "QueryFilterCategoriesResponse",
+					},
+				},
+				EnumValues: map[string][]EnumValueMapping{
+					"CategoryKind": {
+						{Value: "BOOK", TargetValue: "CATEGORY_KIND_BOOK"},
+						{Value: "ELECTRONICS", TargetValue: "CATEGORY_KIND_ELECTRONICS"},
+						{Value: "FURNITURE", TargetValue: "CATEGORY_KIND_FURNITURE"},
+						{Value: "OTHER", TargetValue: "CATEGORY_KIND_OTHER"},
+					},
+				},
+				Fields: map[string]FieldMap{
+					"Query": {
+						"categories": {
+							TargetName: "categories",
+						},
+						"categoriesByKind": {
+							TargetName: "categories_by_kind",
+							ArgumentMappings: map[string]string{
+								"kind": "kind",
+							},
+						},
+						"filterCategories": {
+							TargetName: "filter_categories",
+							ArgumentMappings: map[string]string{
+								"filter": "filter",
+							},
+						},
+					},
+					"Category": {
+						"id": {
+							TargetName: "id",
+						},
+						"name": {
+							TargetName: "name",
+						},
+						"kind": {
+							TargetName: "kind",
+						},
+					},
+					"CategoryFilter": {
+						"category": {
+							TargetName: "category",
+						},
+						"pagination": {
+							TargetName: "pagination",
+						},
+					},
+					"Pagination": {
+						"page": {
+							TargetName: "page",
+						},
+						"perPage": {
+							TargetName: "per_page",
+						},
+					},
+				},
+			}
+
+			compiler, err := NewProtoCompiler(grpctest.MustProtoSchema(t), mapping)
+			if err != nil {
+				t.Fatalf("failed to compile proto: %v", err)
+			}
+
+			// Create the datasource
+			ds, err := NewDataSource(conn, DataSourceConfig{
+				Operation:    &queryDoc,
+				Definition:   &schemaDoc,
+				SubgraphName: "Products",
+				Mapping:      mapping,
+				Compiler:     compiler,
+			})
+			require.NoError(t, err)
+
+			// Execute the query through our datasource
+			output := new(bytes.Buffer)
+			input := fmt.Sprintf(`{"query":%q,"body":%s}`, tc.query, tc.vars)
+			err = ds.Load(context.Background(), []byte(input), output)
+			require.NoError(t, err)
+
+			// Parse the response
+			var resp struct {
+				Data   map[string]interface{} `json:"data"`
+				Errors []struct {
+					Message string `json:"message"`
+				} `json:"errors,omitempty"`
+			}
+
+			err = json.Unmarshal(output.Bytes(), &resp)
+			require.NoError(t, err, "Failed to unmarshal response")
+			require.Empty(t, resp.Errors, "Response should not contain errors")
+			require.NotEmpty(t, resp.Data, "Response should contain data")
+
+			// Run the validation function
+			tc.validate(t, resp.Data)
+		})
+	}
+}
+
+// Test_DataSource_Load_WithTypename tests that __typename fields are correctly included
+// in the response with their static values
+func Test_DataSource_Load_WithTypename(t *testing.T) {
```

Comment:
will typename work for the nested not entity types?

Reply:
Yes